### PR TITLE
Migrate dask-cudf CudfEngine to leverage ArrowDatasetEngine

### LINKED ISSUE #8656: [FEA] Adopt ArrowDatasetEngine for Dask-cuDF read_parquet
**Is your feature request related to a problem? Please describe.**
Dask-cuDF's `read_parquet` code is currently defining a `CudfEngine` class that implicitly inherits from Dask-Dataframe's `ArrowLegacyEngine`.  As discussed in [dask#7871](https://github.com/dask/dask/issues/7871), there are now plans in Arrow to deprecate some legacy-parquet functionality.  Therefore, it is in the best interest of cudf/dask_cudf to adopt the `ArrowDatasetEngine` API.

**Describe the solution you'd like**
I would like to see minimal changes in `CudfEngine` to align with the `ArrowDatasetEngine` API in Dask-Dataframe.

**Describe alternatives you've considered**
Since Dask will eventually **need** to change `ArrowEngine` (which currently points to `ArrowLegacyEngine`) to point to `ArrowDatasetEngine`, the only other option is to wait for the uptream change (and fix anything that breaks).

### PR DESCRIPTION
**Pull request by rjzamora**
Closes #8656 

Addresses the impending deprecation of the ArrowLegacyEngine (which dask-cudf currently depends on), by migrating the `CudfEngine` backend to the newer `ArrowDatasetEngine`.

TODO:

- [x] ~Benchmark/check for any (significant) performance regressions~ (**EDIT**: [pyarrow-deprecations in pyarrow-5](https://github.com/dask/dask/issues/7961) make this migration necessary IMO)

### CODE REVIEW

File: python/dask_cudf/dask_cudf/io/parquet.py [hunk 1]
```diff
@@ -127,18 +132,17 @@ def read_partition(
...
+                categories = partitions[i].keys
 
                 col = as_column(index2).as_frame().repeat(len(df))._data[None]
-                df[name] = build_categorical_column(
-                    categories=categories,
-                    codes=as_column(col.base_data, dtype=col.dtype),
-                    size=col.size,
-                    offset=col.offset,
-                    ordered=False,
+
```
Comment (quasiben) [inline python/dask_cudf/dask_cudf/io/parquet.py:141]: A lot nicer with supported categorical column creation!

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py [hunk 2]
```diff
@@ -455,17 +455,20 @@ def test_create_metadata_file_inconsistent_schema(tmpdir):
     p1 = os.path.join(tmpdir, "part.1.parquet")
     df1.to_parquet(p1, engine="pyarrow")
 
-    with pytest.raises(RuntimeError):
-        # Pyarrow will fail to aggregate metadata
-        # if gather_statistics=True
-        dask_cudf.read_parquet(str(tmpdir), gather_statistics=True,).compute()
+    # New pyarrow-dataset base can handle an incosistent
```
Comment (quasiben) [inline python/dask_cudf/dask_cudf/io/tests/test_parquet.py:458]:
```suggestion
    # New pyarrow-dataset base can handle an inconsistent
```

**Comment by quasiben** [comment]
@rjzamora left one small comment (small typo) . I'll approve now though as i might not be available tomorrow morning

**Review by quasiben — APPROVED** [review]

File: python/dask_cudf/dask_cudf/io/parquet.py (hunk 1 above)
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/io/parquet.py]: I just realized that the new change here was a bit inefficent.  We were creating a column with the partition-based category repeated in every element, and then converting it to a categorical column.  It makes a bit more sense to repeat the index of the partition-based category in every element, and build the categorical column directly.

**Comment by rjzamora** [comment]
Thanks for the review @quasiben !

Note that I did revise a few small things since you approved the PR yesterday.

**Comment by galipremsagar** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**